examples/extensions/add_site_links.py

```diff
@@ -0,0 +1,219 @@
+#!/usr/bin/env python
+# Copyright 2020 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Adds sitelinks to a campaign.
+
+To create a campaign, run add_campaigns.py.
+"""
+
+
+import argparse
+from collections import namedtuple
+import datetime
+import sys
+
+from google.ads.google_ads.client import GoogleAdsClient
+from google.ads.google_ads.errors import GoogleAdsException
+
+
+_DateRange = namedtuple('_DateRange', ['start_datetime', 'end_datetime'])
+_DT_FMT = '%Y-%m-%d %H:%M:%S'
+
+
+def main(client, customer_id, campaign_id):
+    # Create an extension setting.
+    campaign_service = client.get_service('CampaignService', version='v2')
+    campaign_ext_setting_service = client.get_service(
+        'CampaignExtensionSettingService', version='v2')
+
+    campaign_resource_name = campaign_service.campaign_path(
+        customer_id, campaign_id)
+
+    feed_item_resource_names = _create_extension_feed_items(
+        client, customer_id, campaign_resource_name)
+
+    campaign_ext_setting_operation = client.get_type(
+        'CampaignExtensionSettingOperation', version='v2')
+    extension_type_enum = client.get_type(
+        'ExtensionTypeEnum', version='v2')
+
+    campaign_ext_setting = campaign_ext_setting_operation.create
+    campaign_ext_setting.campaign.value = campaign_resource_name
+    campaign_ext_setting.extension_type = (
+        extension_type_enum.SITELINK)
+
+    campaign_ext_setting.extension_feed_items.extend(feed_item_resource_names)
+
+    # Add campaign extension setting with site link feed items.
+    try:
+        response = (
+            campaign_ext_setting_service.mutate_campaign_extension_settings(
+                customer_id, [campaign_ext_setting_operation])
+        )
+    except GoogleAdsException as ex:
+        print(f'Request with ID "{ex.request_id}" failed with status '
+              f'"{ex.error.code().name}" and includes the following errors:')
+        for error in ex.failure.errors:
+            print(f'\tError with message "{error.message}".')
+            if error.location:
+                for field_path_element in error.location.field_path_elements:
+                    print('\t\tOn field: {field_path_element.field_name}')
+        sys.exit(1)
+
+    print('Created CampaignExtensionSetting: '
+          f'"{response.results[0].resource_name}".')
```

Comment:
nit: mixing strings and f-strings

Reply:
Yeah, admittedly, there isn't a whole lot of guidance on the style-correctness of this. Python automagically concatenates these, and one of them doesn't actually need any formatting, so making it an f-string doesn't *seem* necessary, imo. I'll change this if you feel strongly about it.